File: backend/core/batch_status.py
"""
Single source of truth for where a batch — and each file inside it —
sits in the asynchronous processing workflow.

Mirrors the pattern `core/document_types.py` and `core/review_status.py`
already established: one enum shared by the model column
(`database/models.py`), the API contract (`schemas/batch.py`), and the
persistence layer (`database/batch_crud.py`), so all three stay in sync
by construction rather than by convention.

--- Why two enums and not one -----------------------------------------

A batch and a file in it fail differently, and collapsing them would
lose the distinction the whole feature exists to make. A *file* either
worked or it didn't. A *batch* of 100 files where 3 failed is neither
"Completed" nor "Failed" — it is `PARTIALLY_COMPLETED`, and that is the
single most common real outcome once you process documents in bulk. A
shared enum would force that case to be spelled as a lie in one
direction or the other.

--- Terminal states ---------------------------------------------------

`BatchStatus.COMPLETED`, `FAILED` and `PARTIALLY_COMPLETED` are terminal
in the sense that no worker will advance them on its own — but they are
not frozen: retrying failed files moves a batch back to `PROCESSING`
(see `database/batch_crud.py:recompute_batch_progress`). `FAILED` means
every file failed, not that the batch is unrecoverable.
"""
from enum import Enum


class BatchStatus(str, Enum):
    PENDING = "Pending"
    PROCESSING = "Processing"
    COMPLETED = "Completed"
    FAILED = "Failed"
    PARTIALLY_COMPLETED = "Partially Completed"


class BatchFileStatus(str, Enum):
    PENDING = "Pending"
    PROCESSING = "Processing"
    SUCCESS = "Success"
    FAILED = "Failed"


# The states a file can be (re)queued from. A file that is PENDING was
# never picked up; one that FAILED can be retried. PROCESSING and SUCCESS
# are deliberately excluded — that exclusion is the compare-and-set guard
# that stops two workers from processing the same file twice (see
# `database/batch_crud.py:claim_batch_file`).
RETRYABLE_FILE_STATUSES = (BatchFileStatus.PENDING, BatchFileStatus.FAILED)

# Statuses that mean "this batch still has work in flight or waiting",
# used by the progress stream to decide when it can stop polling.
ACTIVE_BATCH_STATUSES = (BatchStatus.PENDING, BatchStatus.PROCESSING)

File: frontend/src/utils/batchStatus.js
/**
 * The batch and batch-file status vocabularies, mirroring
 * `backend/core/batch_status.py`.
 *
 * Written out here rather than derived from whatever the API last
 * returned, for the same reason `utils/reviewStatus.js` is: the status
 * filters have to offer every status, including the ones that happen to
 * have no batches in them right now. A dropdown built from the loaded
 * page can only offer what is already on screen, which makes "show me
 * the failed ones" disappear exactly when there are none visible — the
 * moment it is most useful to be able to ask.
 *
 * --- Why a batch has five states and a file has four ----------------
 *
 * Because a batch of 100 files where 3 failed is neither completed nor
 * failed. `PARTIALLY_COMPLETED` is the single most common real outcome
 * of bulk processing, and the backend models it as its own state rather
 * than forcing it to be spelled as a lie in one direction. The UI has to
 * carry that distinction through, or it undoes the whole point: a
 * partially completed batch is the one an operator has to *do*
 * something about.
 */

export const BATCH_STATUS = {
  PENDING: 'Pending',
  PROCESSING: 'Processing',
  COMPLETED: 'Completed',
  FAILED: 'Failed',
  PARTIALLY_COMPLETED: 'Partially Completed',
}

export const BATCH_STATUSES = Object.values(BATCH_STATUS)

export const BATCH_FILE_STATUS = {
  PENDING: 'Pending',
  PROCESSING: 'Processing',
  SUCCESS: 'Success',
  FAILED: 'Failed',
}

export const BATCH_FILE_STATUSES = Object.values(BATCH_FILE_STATUS)

/**
 * Whether a batch still has work queued or in flight.
 *
 * The single predicate behind three behaviours that must agree: whether
 * to hold the progress stream open, whether the list keeps refreshing
 * itself, and whether the details page shows a live progress bar. If
 * these three ever disagreed about "still running", the UI would show a
 * finished batch with a moving progress bar, or the reverse.
 *
 * Note that a terminal status is not permanent — retrying failed files
 * moves a batch back to `Processing` — which is exactly why this is a
 * question asked of the current status on every update rather than a
 * flag latched once.
 */
export function isBatchActive(status) {
  return status === BATCH_STATUS.PENDING || status === BATCH_STATUS.PROCESSING
}

/** Whether a *file* has reached an outcome. Used to decide if a row can offer a retry. */
export function isFileFinished(status) {
  return status === BATCH_FILE_STATUS.SUCCESS || status === BATCH_FILE_STATUS.FAILED
}

/**
 * "12 of 500 files" style text for a batch's counters.
 *
 * Here rather than inline in the two components that show it, because
 * the pluralisation and the thousands separators are the kind of detail
 * that quietly diverges between a list row and a details header.
 */
export function formatFileCount(count) {
  return `${count.toLocaleString()} file${count === 1 ? '' : 's'}`
}
